src/util/address.rs

```diff
@@ -148,6 +152,171 @@ impl FromStr for AddressType {
     }
 }
 
+/// Version of the WitnessProgram: first byte of `scriptPubkey` in
+/// transaction output for transactions starting with opcodes ranging from 0
+/// to 16 (inclusive).
+///
+/// Structure helps to limit possible version of the witness according to the
+/// specification; if a plain `u8` type will be used instead it will mean that
+/// version > 16, which is incorrect.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum WitnessVersion {
+    /// Current, initial version of Witness Program. Used for P2WPKH and P2WPK
+    /// outputs
+    V0,
+    /// Forthcoming second version of Witness Program, which (most probably)
+    /// will be used for Taproot
+    V1,
+    /// Future (unsupported) version of Witness Program
+    V2,
+    /// Future (unsupported) version of Witness Program
+    V3,
+    /// Future (unsupported) version of Witness Program
+    V4,
+    /// Future (unsupported) version of Witness Program
+    V5,
+    /// Future (unsupported) version of Witness Program
+    V6,
+    /// Future (unsupported) version of Witness Program
+    V7,
+    /// Future (unsupported) version of Witness Program
+    V8,
+    /// Future (unsupported) version of Witness Program
+    V9,
+    /// Future (unsupported) version of Witness Program
+    V10,
+    /// Future (unsupported) version of Witness Program
+    V11,
+    /// Future (unsupported) version of Witness Program
+    V12,
+    /// Future (unsupported) version of Witness Program
+    V13,
+    /// Future (unsupported) version of Witness Program
+    V14,
+    /// Future (unsupported) version of Witness Program
+    V15,
+    /// Future (unsupported) version of Witness Program
+    V16,
+}
+
+impl Display for WitnessVersion {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", *self as u8)
+    }
+}
+
+
+impl FromStr for WitnessVersion {
```

Comment:
Is this actually useful? Where do we directly (de)serialize a witness version? Isn't it always part of some other data structure with its own encoding?

Reply:
I could imagine it being useful somewhere, and it's not much extra code.

Reply:
Yeah, it's not too bad now that I noticed even without it the `UnparsableWitnessVersion` error variant can't be made more expressive.